# EBAC - Regressão II - regressão múltipla

## Tarefa I

#### Previsão de renda

Vamos trabalhar com a base 'previsao_de_renda.csv', que é a base do seu próximo projeto. Vamos usar os recursos que vimos até aqui nesta base.

|variavel|descrição|
|-|-|
|data_ref                | Data de referência de coleta das variáveis |
|index                   | Código de identificação do cliente|
|sexo                    | Sexo do cliente|
|posse_de_veiculo        | Indica se o cliente possui veículo|
|posse_de_imovel         | Indica se o cliente possui imóvel|
|qtd_filhos              | Quantidade de filhos do cliente|
|tipo_renda              | Tipo de renda do cliente|
|educacao                | Grau de instrução do cliente|
|estado_civil            | Estado civil do cliente|
|tipo_residencia         | Tipo de residência do cliente (própria, alugada etc)|
|idade                   | Idade do cliente|
|tempo_emprego           | Tempo no emprego atual|
|qt_pessoas_residencia   | Quantidade de pessoas que moram na residência|
|renda                   | Renda em reais|

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('previsao_de_renda.csv')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             15000 non-null  int64  
 1   data_ref               15000 non-null  object 
 2   id_cliente             15000 non-null  int64  
 3   sexo                   15000 non-null  object 
 4   posse_de_veiculo       15000 non-null  bool   
 5   posse_de_imovel        15000 non-null  bool   
 6   qtd_filhos             15000 non-null  int64  
 7   tipo_renda             15000 non-null  object 
 8   educacao               15000 non-null  object 
 9   estado_civil           15000 non-null  object 
 10  tipo_residencia        15000 non-null  object 
 11  idade                  15000 non-null  int64  
 12  tempo_emprego          12427 non-null  float64
 13  qt_pessoas_residencia  15000 non-null  float64
 14  renda                  15000 non-null  float64
dtypes:

1. Ajuste um modelo para prever log(renda) considerando todas as covariáveis disponíveis.
    - Utilizando os recursos do Patsy, coloque as variáveis qualitativas como *dummies*.
    - Mantenha sempre a categoria mais frequente como casela de referência
    - Avalie os parâmetros e veja se parecem fazer sentido prático.  


2. Remova a variável menos significante e analise:
    - Observe os indicadores que vimos, e avalie se o modelo melhorou ou piorou na sua opinião.
    - Observe os parâmetros e veja se algum se alterou muito.  


3. Siga removendo as variáveis menos significantes, sempre que o *p-value* for menor que 5%. Compare o modelo final com o inicial. Observe os indicadores e conclua se o modelo parece melhor.
    

# 1. Ajuste um modelo para prever log(renda) considerando todas as covariáveis disponíveis.

 * Utilizando os recursos do Patsy, coloque as variáveis qualitativas como dummies.









In [9]:
import pandas as pd
import statsmodels.api as sm
import patsy
import numpy as np

# Carregar os dados
df = pd.read_csv('previsao_de_renda.csv')

# Criar a fórmula com o Patsy, incluindo variáveis qualitativas como dummies
formula = 'np.log(renda) ~ ' + ' + '.join(['C({})'.format(column) for column in df.select_dtypes(include=['object']).columns.difference(['renda'])])

# Criar as matrizes de design usando patsy
y, X = patsy.dmatrices(formula, df, return_type='dataframe')

# Ajustar o modelo de regressão múltipla
model = sm.OLS(y, X)
result = model.fit()

# Imprimir os resultados
print(result.summary())



                            OLS Regression Results                            
Dep. Variable:          np.log(renda)   R-squared:                       0.161
Model:                            OLS   Adj. R-squared:                  0.159
Method:                 Least Squares   F-statistic:                     89.46
Date:                Thu, 21 Dec 2023   Prob (F-statistic):               0.00
Time:                        21:42:34   Log-Likelihood:                -18112.
No. Observations:               15000   AIC:                         3.629e+04
Df Residuals:                   14967   BIC:                         3.654e+04
Df Model:                          32                                         
Covariance Type:            nonrobust                                         
                                          coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------
In

* Utilizando os recursos do Patsy, coloque as variáveis qualitativas como dummies.

In [10]:
import pandas as pd
import statsmodels.api as sm
import patsy
import numpy as np

# Carregar os dados
df = pd.read_csv('previsao_de_renda.csv')

# Identificar variáveis categóricas
categorical_columns = df.select_dtypes(include=['object']).columns

# Criar a fórmula com o Patsy, incluindo variáveis qualitativas como dummies
formula = 'np.log(renda) ~ ' + ' + '.join(['C({0}, Treatment(reference="{1}"))'.format(column, df[column].mode().iloc[0]) for column in categorical_columns.difference(['renda'])])

# Criar as matrizes de design usando patsy
y, X = patsy.dmatrices(formula, df, return_type='dataframe')

# Ajustar o modelo de regressão múltipla
model = sm.OLS(y, X)
result = model.fit()

# Imprimir os resultados
print(result.summary())


                            OLS Regression Results                            
Dep. Variable:          np.log(renda)   R-squared:                       0.161
Model:                            OLS   Adj. R-squared:                  0.159
Method:                 Least Squares   F-statistic:                     89.46
Date:                Thu, 21 Dec 2023   Prob (F-statistic):               0.00
Time:                        21:49:58   Log-Likelihood:                -18112.
No. Observations:               15000   AIC:                         3.629e+04
Df Residuals:                   14967   BIC:                         3.654e+04
Df Model:                          32                                         
Covariance Type:            nonrobust                                         
                                                                            coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------

* Avalie os parâmetros e veja se parecem fazer sentido prático.

O modelo de regressão linear analisado revela insights significativos sobre os determinantes da renda. O intercepto de 7.958 representa a renda média quando todas as variáveis são zero ou na categoria de referência. Os coeficientes para as datas de referência indicam variações na renda em comparação com '2015-01-01', embora alguns não sejam estatisticamente significativos. A educação mostra influência significativa, sendo que ter 'Superior incompleto' está associado a uma diminuição de 0.1521 nas unidades do logaritmo da renda. O estado civil também é relevante, sendo 'União' associado a uma diminuição de 0.0677. Clientes do sexo masculino têm uma renda média exp(0.6976) vezes maior do que os do sexo feminino. Algumas variáveis de tipo de renda e residência têm efeitos significativos. O modelo explica 16.1% da variabilidade na renda, e o teste F é estatisticamente significativo (p = 0.00).

# 2. Remova a variável menos significante e analise:

* Observe os indicadores que vimos, e avalie se o modelo melhorou ou piorou na sua opinião.
* Observe os parâmetros e veja se algum se alterou muito.

In [11]:
import pandas as pd
import statsmodels.api as sm
import patsy
import numpy as np

# Carregar os dados
df = pd.read_csv('previsao_de_renda.csv')

# Remover as variáveis menos significativas
df = df.drop(['Unnamed: 0', 'data_ref'], axis=1)

# Criar a fórmula com o Patsy, incluindo variáveis qualitativas como dummies
formula = 'np.log(renda) ~ ' + ' + '.join(['C({})'.format(column) for column in df.select_dtypes(include=['object']).columns.difference(['renda'])])

# Criar as matrizes de design usando patsy
y, X = patsy.dmatrices(formula, df, return_type='dataframe')

# Ajustar o modelo de regressão múltipla
model = sm.OLS(y, X)
result = model.fit()

# Imprimir os resultados
print(result.summary())


                            OLS Regression Results                            
Dep. Variable:          np.log(renda)   R-squared:                       0.160
Model:                            OLS   Adj. R-squared:                  0.159
Method:                 Least Squares   F-statistic:                     158.0
Date:                Thu, 21 Dec 2023   Prob (F-statistic):               0.00
Time:                        22:11:43   Log-Likelihood:                -18121.
No. Observations:               15000   AIC:                         3.628e+04
Df Residuals:                   14981   BIC:                         3.642e+04
Df Model:                          18                                         
Covariance Type:            nonrobust                                         
                                          coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------
In

**Após remover as variáveis menos significativas, o modelo resultante apresenta as seguintes estatísticas:**

* R-squared: 0.160 (anteriormente 0.161)
* Adj. R-squared: 0.159 (sem alteração)
* F-statistic: 158.0 (anteriormente 89.46)
* Prob (F-statistic): 0.00 (sem alteração)

**Alguns coeficientes específicos também mudaram:**

* C(educacao)[T.Pós graduação]: -0.0951 (anteriormente -0.1012)
* C(educacao)[T.Superior incompleto]: -0.1492 (anteriormente -0.1521)
* C(estado_civil)[T.União]: -0.0695 (anteriormente -0.0677)

# 3. Siga removendo as variáveis menos significantes, sempre que o p-value for menor que 5%. Compare o modelo final com o inicial. Observe os indicadores e conclua se o modelo parece melhor.

In [18]:
import pandas as pd
import statsmodels.api as sm
import patsy
import numpy as np

# Carregar os dados
df = pd.read_csv('previsao_de_renda.csv')

# Remover as variáveis menos significativas
df = df.drop(['Unnamed: 0', 'data_ref' , 'estado_civil' , 'tipo_residencia' ], axis=1 )

# Criar a fórmula com o Patsy, incluindo variáveis qualitativas como dummies
formula = 'np.log(renda) ~ ' + ' + '.join(['C({})'.format(column) for column in df.select_dtypes(include=['object']).columns.difference(['renda'])])

# Criar as matrizes de design usando patsy
y, X = patsy.dmatrices(formula, df, return_type='dataframe')

# Ajustar o modelo de regressão múltipla
model = sm.OLS(y, X)
result = model.fit()

# Imprimir os resultados
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:          np.log(renda)   R-squared:                       0.156
Model:                            OLS   Adj. R-squared:                  0.156
Method:                 Least Squares   F-statistic:                     308.7
Date:                Thu, 21 Dec 2023   Prob (F-statistic):               0.00
Time:                        22:31:01   Log-Likelihood:                -18149.
No. Observations:               15000   AIC:                         3.632e+04
Df Residuals:                   14990   BIC:                         3.639e+04
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
                                         coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------
Inte

## Consideraçoes finais:

### Após a remoção das variáveis ['Unnamed: 0', 'data_ref', 'estado_civil', 'tipo_residencia'], o modelo de regressão linear parece ter passado por algumas mudanças significativas. Aqui estão algumas considerações finais com base nos resultados analizados:

1. **Redução de Variáveis:** A remoção dessas variáveis menos significativas levou a um modelo mais simplificado, com menos preditores.

2. **Estatísticas Globais do Modelo:** O valor do F-statistic aumentou significativamente para 308.7, indicando que o modelo geral é estatisticamente significativo. Isso sugere que pelo menos uma das variáveis independentes tem um efeito significativo na variável dependente.

3. **R-squared:** O coeficiente de determinação (R-squared) aumentou para 0.156, indicando que o modelo explica aproximadamente 15.6% da variabilidade na variável dependente. Isso representa uma melhoria em relação ao modelo anterior.

4. **Coeficientes Individuais:** Os coeficientes para as variáveis remanescentes indicam a direção e magnitude da relação com a variável dependente. Por exemplo, o coeficiente positivo para 'C(sexo)[T.M]' sugere um aumento na renda para clientes do sexo masculino em comparação com o sexo feminino.

5. **P-valores:** Os p-valores associados aos coeficientes indicam a significância estatística de cada variável. Valores-p baixos (geralmente abaixo de 0.05) são considerados estatisticamente significativos.








